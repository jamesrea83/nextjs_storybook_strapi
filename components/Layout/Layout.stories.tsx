import { Meta, StoryObj } from '@storybook/react';
import { expect } from '@storybook/jest';
import { screen, userEvent } from '@storybook/testing-library';
import { Layout } from '.';

export default {
	title: 'Content/Layout',
	component: Layout,
} as Meta<typeof Layout>;

type Story = StoryObj<typeof Layout>;

export const Primary: Story = {
	args: {
		children: 'Main Page',
	},
};
