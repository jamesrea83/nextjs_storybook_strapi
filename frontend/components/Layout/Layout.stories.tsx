import { Meta, StoryObj } from '@storybook/react';
import { Layout } from '.';

export default {
	title: 'Content/Layout',
	component: Layout,
} as Meta<typeof Layout>;

type Story = StoryObj<typeof Layout>;

const child = (
	<>
		<h1>Main article area</h1>
		<p>
			Lorem ipsum dolor sit amet. Eum quaerat odio cum soluta quas aut sunt
			repellat et aliquid laboriosam. Aut corrupti excepturi est consequatur
			nostrum aut quos facere. Sit nobis voluptatum non veniam fuga est ducimus
			mollitia vel perspiciatis mollitia aut commodi praesentium vel officiis
			repellendus.
		</p>
	</>
);

export const Primary: Story = {
	args: {
		children: child,
	},
};
