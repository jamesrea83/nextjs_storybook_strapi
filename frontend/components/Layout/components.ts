import styled from '@emotion/styled';

import { Logo } from '@/components/Logo';
import { Input } from '@/components/Input';
import { StyledLink } from '@/components/StyledLink';

export const Wrapper = styled.div`
	display: grid;
	gap: 0.1rem;
	color: ${({ theme }) => theme.font.regular};
	background-color: ${({ theme }) => theme.background};
	padding: 0.5rem;
	grid-template-areas:
		'header nav'
		'search search'
		'content content'
		'footer footer';
	nav {
		justify-content: flex-end;
		gap: 5vmin;
	}
	@media (min-width: 500px) {
		grid-template-columns: 1fr 3fr;
	}
	@media (min-width: 960px) {
		grid-template-columns: 1fr 4fr 2fr;
		grid-template-areas:
			'header		search 	nav'
			'content 	content	content'
			'footer 	footer	footer';
	}
`;

export const LogoLink = styled(StyledLink)`
	padding-right: 1vw;
`;

export const StyledLogo = styled(Logo)`
	grid-area: header;
	display: flex;
	align-items: center;
	height: 4rem;
	& .logo_full {
		display: none;
	}
	@media (min-width: 560px) {
		& .logo_short {
			display: none;
		}
		& .logo_full {
			display: inline;
		}
	}
`;

export const MainNav = styled.nav`
	grid-area: nav;
	display: flex;
	align-items: center;
	justify-content: space-around;
	margin: 0 2vmin;
`;

export const SearchInput = styled(Input)`
	grid-area: search;
	width: 100%;
	height: 4rem;
`;

export const Content = styled.main`
	grid-area: content;
	min-height: 84vh;
	margin-top: 1rem;
`;

export const Footer = styled.footer`
	grid-area: footer;
	display: flex;
	align-items: center;
	justify-content: space-around;
	height: 5rem;
`;
