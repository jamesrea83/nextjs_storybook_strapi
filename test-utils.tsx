import { FC, ReactElement, ReactNode } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ThemeProvider } from '@emotion/react';

import { Themes } from './styles/themes';

interface Props {
	children: ReactNode;
}

const Wrapper = ({ children }: Props) => (
	<ThemeProvider theme={Themes.light}>{children}</ThemeProvider>
);

const customRender = (ui: ReactElement, options?: RenderOptions) =>
	render(ui, { wrapper: Wrapper, ...options });

export * from '@testing-library/react';

export { customRender as render };
