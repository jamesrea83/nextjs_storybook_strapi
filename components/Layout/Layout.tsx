import { FC, ReactNode } from 'react';
import Link from 'next/link';
import styled from '@emotion/styled';

import { Logo } from '@/components/Logo';
import { Input } from '@/components/Input';
import { IconButton } from '@/components/IconButton';
import { transition } from '@/components/styles';

const Wrapper = styled.div`
	display: grid;
	gap: 0.1rem;
	color: ${({ theme }) => theme.font.regular};
	background-color: ${({ theme }) => theme.background};
	padding: 0.5rem;
	grid-template-areas:
		'header'
		'nav'
		'search'
		'content'
		'sidebar'
		'footer';
	${transition()}
	@media (min-width: 500px) {
		grid-template-columns: 1fr 3fr;
		grid-template-areas:
			'header 	search'
			'nav 		nav'
			'content 	content'
			'footer 	footer';
		nav {
			flex-direction: row;
			justify-content: space-between;
		}
	}
`;

const StyledLogo = styled(Logo)`
	grid-area: header;
`;

const MainNav = styled.nav`
	grid-area: nav;
	display: flex;
	align-items: center;
	justify-content: space-around;
	padding: 0.5rem;
	a {
		cursor: pointer;
		color: ${({ theme }) => theme.font.regular};
		&:hover {
			opacity: 0.7;
		}
	}
`;

const SearchInput = styled(Input)`
	grid-area: search;
	width: 100%;
	height: 4rem;
`;

const Content = styled.main`
	grid-area: content;
`;

const Footer = styled.footer`
	grid-area: footer;
`;

type LayoutProps = {
	children: ReactNode;
};

export const Layout: FC<LayoutProps> = ({ children }) => {
	return (
		<Wrapper>
			<StyledLogo size={3}>C8X</StyledLogo>
			<MainNav>
				<Link href="/all">All</Link>
				<Link href="/news">News</Link>
				<IconButton name="Moon" size={1} onClick={() => null} />
			</MainNav>
			<SearchInput icon="Search" placeholder="Search" onChange={() => null} />
			<Content>{children}</Content>
			<Footer>Jimbob</Footer>
		</Wrapper>
	);
};
