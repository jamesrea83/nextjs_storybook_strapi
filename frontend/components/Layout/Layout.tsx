import { FC, ReactElement, useState, useEffect, useLayoutEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { ThemeProvider } from '@emotion/react';

import { AppDispatch, RootState } from '@/store';
import { login, selectUser } from '@/services/userSlice';

import { IconButton } from '@/components/IconButton';
import { StyledLink } from '@/components/StyledLink';
import { Themes } from '@/styles/themes';
import {
	Wrapper,
	LogoLink,
	StyledLogo,
	MainNav,
	SearchInput,
	Content,
	Footer,
} from './components';

type LayoutProps = {
	children?: ReactElement;
};

const useCustomLayoutEffect =
	typeof window !== 'undefined' ? useLayoutEffect : useEffect;

export const Layout: FC<LayoutProps> = ({ children }) => {
	const { username } = useSelector<RootState, RootState['user']>(selectUser);
	const [isDark, setIsDark] = useState(false);
	const dispatch = useDispatch<AppDispatch>();
	const toggleDark = () => {
		localStorage.setItem('theme', isDark ? 'light' : 'dark');
		setIsDark(!isDark);
	};

	useCustomLayoutEffect(() => {
		dispatch(login());
		const isDark = Boolean(localStorage.getItem('theme') === 'dark');
		setIsDark(
			window.matchMedia('prefers-color-scheme: dark').matches || isDark
		);
	}, []);

	const theme = Themes[isDark ? 'dark' : 'light'];

	return (
		<ThemeProvider theme={theme}>
			<Wrapper>
				<LogoLink href="/">
					<StyledLogo size={3}>
						<span className="logo_short">C8X</span>
						<span className="logo_full">CoursesBox</span>
					</StyledLogo>
				</LogoLink>
				<MainNav>
					<StyledLink href="/all">All</StyledLink>
					<StyledLink href={username ? '/user' : '/login'}>
						<IconButton name={username ? 'User' : 'Login'} size={1} />
					</StyledLink>
					<IconButton
						name={isDark ? 'Sun' : 'Moon'}
						size={1}
						onClick={toggleDark}
					/>
				</MainNav>
				<SearchInput icon="Search" placeholder="Search" onChange={() => null} />
				<Content>{children}</Content>
				<Footer>Footer</Footer>
			</Wrapper>
		</ThemeProvider>
	);
};
