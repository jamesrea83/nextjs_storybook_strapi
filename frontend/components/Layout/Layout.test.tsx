import { render, screen, waitFor } from '@/test-utils';
import userEvent from '@testing-library/user-event';
import { act } from 'react-dom/test-utils';
import { Layout } from './Layout';

describe('Layout tests', () => {
	localStorage.setItem('theme', 'light');
	const child = (
		<>
			<h1>Main article area</h1>
			<p>
				orem ipsum dolor sit amet. Eum quaerat odio cum soluta quas aut sunt
				repellat et aliquid laboriosam. Aut corrupti excepturi est consequatur
				nostrum aut quos facere. Sit nobis voluptatum non veniam fuga est
				ducimus mollitia vel perspiciatis mollitia aut commodi praesentium vel
				officiis repellendus.
			</p>
		</>
	);

	it('should render', async () => {
		const { asFragment } = render(<Layout>{child}</Layout>);

		await waitFor(() => {
			expect(asFragment()).toMatchSnapshot();
		});
	});

	it('should toggle theme', async () => {
		(window.matchMedia as jest.Mock).mockReturnValue({ matches: true });

		render(<Layout>{child}</Layout>);

		const themeToggler = screen.getByRole('button', { name: 'Sun' });
		await waitFor(() => {
			expect(themeToggler).toBeInTheDocument();
		});

		await act(() => userEvent.click(themeToggler));

		await waitFor(() => {
			expect(screen.getByRole('button', { name: 'Moon' })).toBeInTheDocument();
		});
	});
});
