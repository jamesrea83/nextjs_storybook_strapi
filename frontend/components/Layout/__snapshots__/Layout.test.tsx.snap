// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout tests should render 1`] = `
<DocumentFragment>
  <div
    class="css-7ldzj2"
  >
    <a
      class="css-o4zmmn"
      href="/"
    >
      <header
        class="css-1o718mr"
        size="3"
      >
        <span
          class="logo_short"
        >
          C8X
        </span>
        <span
          class="logo_full"
        >
          CoursesBox
        </span>
      </header>
    </a>
    <nav
      class="css-181tzcb"
    >
      <a
        class="css-nzbn0p"
        href="/all"
      >
        All
      </a>
      <a
        class="css-nzbn0p"
        href="/news"
      >
        News
      </a>
      <button
        class="css-636gg9"
        title="Moon"
      >
        <svg
          aria-label="Moon"
          class="css-2gue7k"
          fill="currentColor"
          height="1rem"
          role="img"
          viewBox="0 0 16 16"
          width="1rem"
        >
          <path
            d="M6 .278a.768.768 0 01.08.858 7.208 7.208 0 00-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 01.81.316.733.733 0 01-.031.893A8.349 8.349 0 018.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 016 .278zM4.858 1.311A7.269 7.269 0 001.025 7.71c0 4.02 3.279 7.276 7.319 7.276a7.316 7.316 0 005.205-2.162c-.337.042-.68.063-1.029.063-4.61 0-8.343-3.714-8.343-8.29 0-1.167.242-2.278.681-3.286z"
          />
        </svg>
      </button>
    </nav>
    <label
      class="css-1f2mqsx"
      height="7"
      width="20"
    >
      <span
        class="css-porbx7"
      />
      <div
        class="css-bnug8b"
      >
        <input
          class="css-58wzc"
          placeholder="Search"
        />
        <svg
          aria-label="Search"
          class="css-jp31fy"
          fill="currentColor"
          height="2rem"
          role="img"
          viewBox="0 0 1024 1024"
          width="2rem"
        >
          <path
            d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
          />
        </svg>
      </div>
      <span
        class="css-tx5ech"
      />
    </label>
    <main
      class="css-6z89iq"
    >
      <h1>
        Main article area
      </h1>
      <p>
        orem ipsum dolor sit amet. Eum quaerat odio cum soluta quas aut sunt repellat et aliquid laboriosam. Aut corrupti excepturi est consequatur nostrum aut quos facere. Sit nobis voluptatum non veniam fuga est ducimus mollitia vel perspiciatis mollitia aut commodi praesentium vel officiis repellendus.
      </p>
    </main>
    <footer
      class="css-1b0zqog"
    >
      Footer
    </footer>
  </div>
</DocumentFragment>
`;
